web_interface/astpp/application/controllers/currencyupdate.php

```diff
@@ -28,25 +31,72 @@ function __construct() {
-		$last_updated = $currencyData->date;
-			foreach ($currencyData as $currencykey => $currencyvalue) {			
-				foreach ($currencyvalue as $key => $value) {
-					 $sql = "UPDATE currency SET currencyrate = '".$value."',last_updated = '".$last_updated."' WHERE currency = '" . $key . "'\n";
-					$this->db->query ( $sql );
-				}			
-			}
-		$updatebasecurrency = "UPDATE currency SET currencyrate = '1.000',last_updated = '" .$last_updated."' WHERE currency = '" . Common_model::$global_config ['system_config'] ['base_currency'] . "'";
-		$this->db->query ( $updatebasecurrency );
-$this->session->set_flashdata ( "astpp_errormsg", "Currency exchange rates successfully updated." );
-		redirect ( base_url () . "/systems/currency_list/" );
-		exit;
-	}
+
+    function update_currency() {
+        $apiKey = Common_model::$global_config ['system_config'] ['currency_conv_api_key'];
+        if (Common_model::$global_config ['system_config'] ['currency_conv_api_key'] != ""
+            && Common_model::$global_config ['system_config'] ['currency_conv_api_key'] != "YourKeyGoesHere"
+        ) {
+            $baseCurrency = Common_model::$global_config ['system_config'] ['base_currency'];
+            $url = "https://forex.1forge.com/1.0.3/symbols?api_key=".$apiKey."";
+            $symbolsData = $this->curl_response ( $url );
+            $symbolsData = json_decode($symbolsData);
+            $symbolString = "" ;
+            foreach ($symbolsData as $symbol) {
+                if ( substr($symbol,0, 3) == $baseCurrency ) {
+                    $curQry = "select currency from astpp.currency WHERE currency REGEXP '[A-Z][A-Z][A-Z]';";
+                    $curData = $this->db->query($curQry);
+                    if ($curData->num_rows() > 0) {
+                        $currency_arr = $curData->result_array();
+                        foreach ( $currency_arr  as $currency ) {
+                            foreach ($currency  as $key => $value ) {
+                                if (strtolower($value) == strtolower(substr($symbol, 3, 3))) {
+                                    if ($symbolString == "") {
+                                        $symbolString = $symbol;
+                                    } else {
+                                        $symbolString = $symbolString . "," . $symbol;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            $url = "https://forex.1forge.com/1.0.3/quotes?pairs=".$symbolString."&api_key=".$apiKey."";
+            $quotesData = $this->curl_response ( $url );
+            $quotesData = json_decode($quotesData);
+            $debugstring = "";
+            foreach ($quotesData as $quote) {
+                $debugstring = $debugstring . "A," ;
+                if (isset(Common_model::$global_config ['system_config'] ['currency_conv_loss_pct'])
```

Comment:
Hi @AlexHeylin 

Thanks for the PR. 

Would you please explain the use of currency_conv_lost_pct?

Reply:
When changing currency it's usual for there to be fees involved.  For example 5% bank fee - in this case you'd set currency_conv_lost_pct to 5 (in Configuration > Settings > Global > Currency Conversion Loss Percentage) and when the system imports the market rate, it will allow for this 5% loss and put the rate you actually get into the DB.  This helps ensure the system calculates conversion more accurately.  Say the market rate is 2 USD = 1 GBP, and base currency is GBP, if I set currency_conv_lost_pct to 5 the rate in the DB will be 1.9 USD = 1 GBP.

If you don't want to allow for any loss (the way it used to be) then just set Currency Conversion Loss Percentage = 0 and it will use the rate directly from the API.

The two new system variables need to be created by importing the SQL also in the PR. If use manually runs currency update without setting the API key, if will return error message saying to set it in Configuration > Settings > Global > API key for currency rate import from 1forge.com

Please don't merge this yet - I've seen a couple of improvements (remove debug / check DB write worked / log to disk) I should make. Also this is failing on one of my systems and I've not yet found out why.  If I make more changes to the code in the same branch, do I need to do a new PR or just let you know this one is OK to process?
Thanks @smrdoshi

Reply:
Thanks for explanation @AlexHeylin 

I got the idea. Let's do it. 
Once you will be done all the modifications in your branch, you will need to send new PR.